### ISSUE TITLE
Clean up configure script

### ISSUE THREAD

**Issue by darnir** [neomutt/neomutt#58]
Taking a look at the build system, I saw lots of issues with the configure.ac script that can be cleaned up.

1. The various configure options often use the wrong terminology. --enable options should enable features in the codebase while --with options should enable support using 3rd party libraries. Hence, it should be --with-gpgme and --with-pgp.
2. There is no need to test for `x$enableval = yes` in the AC_ARG_ENABLE macros. In the 3rd argument, you can simply set the variable to the value of $enableval. Take a look at this configure script for ideas: https://github.com/rockdaboot/wget2/blob/master/configure.ac
Tim has spent considerable time making sure the configure.ac script is perfect and we've had enough long discussions about the entire autotools tree.
3. We can have configure print a summary of build options at the end. This is often useful debug information, especially for devs. Again, take a look at the above cofngure script for what it does.

I don't mind working on these points, but I'll do it only after discussion and if I know that the patches will be accepted. It would be nice to upstream these some of these patches as well for the cases where they are similar.

**Comment by flatcap** [comment]
Wow!  You really like to tackle the big issues.

Two things I should point out:
- Most of the configure code for _NeoMutt_ was cut'n'pasted, from Mutt, without much understanding.
- Recent changes were made to synchronise configure.ac with upstream Mutt.

Things you can help me with:
- Compress - It's a small start, but if you make this a perfect example, I'll cut'n'paste it for new features.
- Configure summary - I like the idea, but I think it should default to off (perhaps: --show-summary)
- Fmemopen - This is a feature of glibc (HAVE_FMEMOPEN).  If the user has glibc, they get this feature.  I'd like it to be optional.

> I don't mind working on these points, but I'll do it only after discussion and if I know that the patches will be accepted.

That's perfectly reasonable.

> It would be nice to upstream these some of these patches as well

I'm happy to accept your expertise, here, but I think that upstream might be the place to start.
If you're not already on the Mutt Dev mailing list, [here are the instructions](http://www.mutt.org/mail-lists.html).

I suggest posting your concerns and perhaps show the difference that tidying `$enableval` would make.

Fixing --enable/--with is going to be a tough sell because it will immediately break all the downstream distros' builds.  How much duplicate code would be involved in supporting **both** (temporarily)?

**Comment by darnir** [comment]
>
> Two things I should point out:
>
>    - Most of the configure code for *NeoMutt* was cut'n'pasted, from
>    Mutt, without much understanding.
>    - Recent changes were made to synchronise configure.ac with upstream
>    Mutt.
>
> I figured that a lot of the build tools were simply taken from upstream
Mutt.

However, can you tell me what recent changes have you made to synchronize
with upstream? So I know where we stand compared to them.
I hate working with upstream Mutt because its based on Mercurial. I've had
enough pains trying to maintain the sidebar patch for Arch Linux before
moving on to NeoMutt.

>
>
> Things you can help me with:
>
>    - Compress - It's a small start, but if you make this a perfect
>    example, I'll cut'n'paste it for new features.
>    - Configure summary - I like the idea, but I think it should default
>    to off (perhaps: --show-summary)
>    - Fmemopen - This is a feature of glibc (HAVE_FMEMOPEN). If the user
>    has glibc, they get this feature. I'd like it to be optional.
>
> * Compress: This is easy. I'll write up a nice example for one of the
options. That then simply needs to be replicated for all the others. It's
fairly simple to do this
* Configure Summary: The whole point of a configure summary is that it is
always visible. It helps developers see at a glance how the software is
going to be compiled. What features are enabled and what compiler flags are
set. Take a look at the configure output of Wget or Wget2 and you'll see
what I mean. There's no point in keeping it optional.
* Fmemopen: I know about this. But that is something that would come later
after all the cleanups to the configure code. This is a new feature that
should not be mixed with the cleanup

> I'm happy to accept your expertise, here, but I think that upstream might
> be the place to start.
> If you're not already on the Mutt Dev mailing list, here are the
> instructions <http://www.mutt.org/mail-lists.html>.
>
Upstream is a little annoying to work with because it Hg. I'll take a look
and send an email maybe tonight.

> I suggest posting your concerns and perhaps show the difference that
> tidying $enableval would make.
>
I'll do that. Let's see about that. However, would you be willing to merge
these changes into NeoMutt irrespective of upstream? They can always cherry
pick it from here if they want.

> Fixing --enable/--with is going to be a tough sell because it will
> immediately break all the downstream distros' builds. How much duplicate
> code would be involved in supporting *both* (temporarily)?
>
They will probably not accept such a change. Even though it would be nice.
It makes no sense to duplicate the code in configure.ac. That would only
cause a lot of confusion. And going by my experience working with GNU Wget,
it is impossible to remove added code. You will end up supporting both
which is a nightmare. Must pick a side, can't sit on the fence for this one


-- 
Thanking You,
Darshit Shah

**Comment by flatcap** [comment]
> what recent changes have you made to synchronize with upstream?

Upstream fixed how source/object names were handled: 05a8e75 (look at Makefile.am, configure.ac).
I made the same changes to Compress: 53d7c9c.

> I hate working with upstream Mutt because it's based on Mercurial

I feel your pain :-)

I have a clone of Mutt's hg repo which I merge into my master git repo.  That's all I see of Mercurial.

> > Compress:

> This is easy. I'll write up a nice example for one of the options.

Thanks.

> > * Configure Summary: 

> The whole point of a configure summary is that it is always visible.
> There's no point in keeping it optional.

OK, fair enough.  Please create it.

> > * Fmemopen:

> I know about this. But that is something that would come later

OK.  It's very low priority.

> > I suggest posting your concerns and perhaps show the difference that
> >  tidying $enableval would make.

> I'll do that.
> However, would you be willing to merge these changes into NeoMutt irrespective of upstream?

Yes, I am.

Makefile.am and configure.ac rarely change so synchronisation won't be a problem.

> They can always cherry pick it from here if they want.

Hehe, we'll see how persuasive you can be on the dev mailing list :-)

> > Fixing --enable/--with

> They will probably not accept such a change.

I think so, too.

> It makes no sense to duplicate the code in configure.ac.
> it is impossible to remove added code.

Just a crazy thought...

**Comment by flatcap** [comment]
I think you should have enough privileges to create a new branch in NeoMutt.
I suggest [devel/configure] branched from [neomutt].

Ultimately, your changes will need to be split up into the various feature branches, then merged into NeoMutt.

**Comment by darnir** [comment]
No, simply using a Makefile is nearly impossible for such a large project. We need a configure file to make sure that all the required features exist on the system and to accordingly create the Makefile.

The process goes like this:
1. Developer writes `configure.ac` and states his assumptions about the system in it. This includes required libraries and also options for which program features are enabled or disabled.
2. Developer writes a `Makefile.am` and provides a high level description of how the program is to be compiled. This includes a list of source files, how the distribution is created and other things.
3. Now comes the complex part. I'll gloss over details, but autotools will read the `configure.ac` file and generate a shell script called `./configure`. This script is significantly longer than the developer written file and does hundreds of sanity checks on the system. When executed, it reads the user's configure options which define what program features to enable or disable and identifies what library resides where. The script will then read the developer provided `Makefile.am` and churn out a `Makefile`.
4. The Makefile is specific to the configure options provided by the user and must be regenerated everytime they change.

Since the configure script does hundreds (maybe thousands) of sanity checks, it allows us, the developers to not bother with them. It will identify whether the compiler works, and what flags it accepts. How it handles dependencies, etc, and create the Makefile accordingly.
While for small projects just a Makefile may suffice, it definitely won't for us. There is a good reason that these are long standing full projects for so many years, It takes so much time to ensure that your build works everywhere. We really do not want to do this ourselves.

As far as how to use it goes, if you ever touch `configure.ac`, then you should simply execute `autoreconf -i` immediately to recreate the `configure` file. If you touch `Makefile.am`, then remember to re-run `./configure`with all the options. I generally define my configure command for each project as a bash alias make it easy for myself.

Autotools is a big beast that few people understand. It is unweildy, but it does its job well. And we don't have a lot of better alternatives. Cmake probably. But that has its own set of issues. Grass is always greener on the other side.

**Comment by darnir** [comment]
> I think you should have enough privileges to create a new branch in NeoMutt.
> I suggest [devel/configure] branched from [neomutt].
> 
I've started working on a local branch that I'm pushing to a forked repository on my account. I can send in PRs from there or create a branch [devel/configure], either way works with me. Let me know how you want it

> Ultimately, your changes will need to be split up into the various feature branches, then merged into NeoMutt.
True. 

Currently, you can take a look at what I propose here: https://github.com/darnir/neomutt/commit/578a0b397c73acbc8abb2fb95d0f27bc957db2c4 and https://github.com/darnir/neomutt/commit/02d375cbd3fbdc1ac823e3a4b132adb6d011903a

I've done the same for the GPGME option as well, but that will require a little more cleaning that I intend to work on later. Higher level stuff first, then go deeper into the polishing

**Comment by pickfire** [comment]
@darnir Thanks a lot for the info, there might be a bit misunderstanding for me since some communities hate autotools, maybe now I know why some still uses autotools.

**Comment by darnir** [comment]
@flatcap: Quick question, was compressed folders support working fine earlier? Because I think I spotted a bug in the configure script which might cause it to not compile properly. I've cleaned it up in this commit: https://github.com/darnir/neomutt/commit/78a595b399d7f0256cdee90780bc5eb04e8d7595

Do test it and let me know.

**Comment by flatcap** [comment]
@darnir 

Thanks for the summary of autoconf.  I didn't learn anything new, but it's reassuring to know that I understand.

> I've started working on a local branch

Whatever's easiest for you.

> I can send in PRs from there or create a branch [devel/configure]

No need to create another branch.  Either:
- Create a PR, and I'll split and merge the changes
- Or, if you want a challenge, I'll explain the process and you can try

Changes:
- darnir@578a0b3 - Summary: Lovely
- darnir@02d375c - PGP - Very good.  That's the sort of thing that'll get the Mutt devs listening to you.
- darnir@78a595b - Compress: Lovely

> was compressed folders support working fine earlier

Yes, maybe :-)
I tested that it compiled with and without --enable-compressed.
I don't remember testing with an actual compressed file, though :-(

If you want to propose an automated testing framework, [start here](https://github.com/neomutt/neomutt/issues/new)!

**Comment by flatcap** [comment]
@pickfire 

> Is there a way to use only a Makefile and a config.h to configure mutt?

There is a different way to work, but it's extreme in the other direction to autoconf.
Ages ago, I created a "min-build" for NeoMutt.  Often when I'm exploring and learning about Mutt, I like to get rid of all the non-code bits.  I've turned it into a script.  It _may_ be of use to you.

The script deletes ALL non-code files, flattens imap/* into the current directory and creates a new config.h and Makefile (that can be read by a human).  By default everything's enabled.

```
git clone https://github.com/neomutt/neomutt
git clone https://github.com/neomutt/scripts
cd neomutt
../scripts/make-min/make-min-build.sh
make
```

**Comment by pickfire** [comment]
@flatcap, interesting with the option `--enable-everything`, that's what I am looking for. (instead of `./configure --help` and add everything)

**Comment by flatcap** [comment]
@pickfire 
You would have to make some assumptions, e.g. which hcache backend to use.
It would probably mean a big change to configure to make it work (without defaulting everything to 'on').
Like @darnir, I have an alias to configure what I want.

**Comment by darnir** [comment]
Can I open a pull request here for these changes?

**Comment by flatcap** [comment]
If you're planning to work on them, yes.

**Comment by darnir** [comment]
I already showed the working commits. Which branch / repo should I open a request against?

**Comment by flatcap** [comment]
If you're happy the configure changes are solid, then please commit each one to the respective feature branch.  Any global changes, e.g. options summary, please create a devel branch from neomutt.

**Comment by c-alpha** [comment]
In any case **(and in the short term?)** the autoreconf-ed files should be added to the repo. As is now, one needs autoconf and autotools installed to be able to build and deploy. A bit of an overkill IMO if you're not a dev working on the build scheme. A high hurdle for users, esp. in the absence of binary packages.

**Comment by flatcap** [comment]
Welcome to the discussion, @c-alpha.
I'll start at the end:

> esp. in the absence of binary packages.

**We** don't provide any binaries, but I'm helping several distros who do.
- Fedora
- Gentoo
- Arch
- HomeBrew (OSX)
- Suse
- Debian
- OpenBSD
- FreeBSD

**For the casual user**, I'd prefer they used a packaged version of NeoMutt.
It's also less of a support burden for a small project.

**For developer types**, who already have a build environment, autoconf is just one more package to install.  Another hurdle, certainly, but not a high one.  I'm working on some build instructions, to make things easier for them.

**For me**, however, it's an extra step I need to remember **every** time I merge something into NeoMutt -- "Have I remember to regenerate the autoconf files?"

I'll certainly give it some more thought.

**Comment by c-alpha** [comment]
It was not my intention to ask for binary packages. Apologies if that was your impression.

> For me, however, it's an extra step I need to remember every time I merge something into NeoMutt -- "Have I remember to regenerate the autoconf files?"

You could craft a small makefile that checks these dependencies, autoreonf-s if necessary, and commits and pushes to github, and creates a release tag for github. Could e.g. be invoked by something similar to:
`make -f release.make git-milestone REL="x.y.z"`
No warranties if github releases can be created from the command line, however; never tried it. But even if its not, the rest may still seem worth automating? Just my two cents anyway.

**Comment by c-alpha** [comment]
I'm not a big fan of package installation systems for OS X (too many of them, each one has one thing but not the other). So I set out to build from source. For anyone else who wants to build from source for the first time (e.g. new developers), I thought I should share my findings and save everyone a couple of hours figuring out what requires what. The diagram below does not show all dependencies, because I wanted to indicate what needs to be built first. If a higher-level package requires another one that a lower-level one already required, I didn't add a line. Hence, if you go from bottom up, you should be fine. I am also indicating some key `configure` options. No warranties for completeness. Hope this can be helpful.
![neomutt-missing-dependencies-os_x](https://cloud.githubusercontent.com/assets/2213269/17642820/704119c8-6155-11e6-8a76-d5375f4e5cc6.png)

**Comment by flatcap** [comment]
> It was not my intention to ask for binary packages

No, no.  No problem. 

> You could craft a small makefile...

That's sounds ingenious.  That would help for the actual release, but neomutt (the branch) gets quite a lot of merges between releases (both Arch and HomeBrew build from git).  I'll still have to take care not to break their build.

Aside: The NeoMutt source branches are like *a maze of twisty little passages, all alike*.

> if github releases can be created

Yes, they can.  GitHub's API is pretty extensive.
Every release I try to automate a little more than before.

Thanks.  More for me to think about.

**Comment by flatcap** [comment]
Oh that's a **very** nice diagram.

Have you worked out the dependencies by hand, or do you have some clever tool?

**Comment by c-alpha** [comment]
> Oh that's a very nice diagram.

Thanks! Glad you find it useful.

> Have you worked out the dependencies by hand, or do you have some clever tool?

By hand...

**Comment by pickfire** [comment]
I think it would be nice to have a `--with-features=huge` for those who doesn't want to read the 300+ lines INSTALL file with all the features people want (like neovim).

**Comment by flatcap** [comment]
@c-alpha 
> I also had to apply a small number of patches
> I can ... upload a .tar.gz so you can diff against the vanilla packages

Yes, please.

**Comment by flatcap** [comment]
@pickfire 
> nice to have a --with-features=huge

You're not the first to mention a `--enable-everything` option.
It would mean altering every branch's build, but I'm starting to like the idea.

All the features, plus one of the hcache options if dependencies are matched.

> 300+ lines INSTALL file

Which hasn't been changed in 7 years!

**Comment by c-alpha** [comment]
>> I also had to apply a small number of patches
>> I can ... upload a .tar.gz so you can diff against the vanilla packages
> 
> Yes, please.

Please get it from http://ooz7eiwa7u.condition-alpha.com/neomutt-osx-buildenv.tar.gz

The tarball is some 111MB, and unpacked it grows to some 389MB. Inside, you will find the vanilla packages as tarballs (so you don't have to find and download them again), and the unpacked and modified by me sources in directories.

To compare you will have to unpack the tarballs to somewhere else and `diff` against my directories. I did do `make distclean` everywhere, so hope the `diff` will not be too noisy (_fingers crossed_).

**Comment by flatcap** [comment]
thanks, @c-alpha

**Comment by c-alpha** [comment]
Just discovered one more oddity: when building `talloc` in `/some/build/dir`, the resulting library contains a reference to self, i.e. to `/some/build/dir/.../libtalloc.dylib` (shared libs are called dylib on OS X). This causes anything (in this case mutt) that has been built against `talloc` to fail launching after you have removed `/some/build/dir`. I am still struggling to figure out how to fix this as `talloc` uses some python-based build scheme I have never worked with before.

If anyone who has experience with `talloc`'s (or Samba's for that matter) build system, any help would be appreciated.
